core/toolbox/toolbox.js

```diff
@@ -0,0 +1,459 @@
+ *     blocks.
+ * @constructor
+ * @implements {Blockly.IBlocklyActionable}
+ * @implements {Blockly.IDeleteArea}
+ * @implements {Blockly.IStyleable}
+ * @implements {Blockly.IToolbox}
+ */
+Blockly.Toolbox = function(workspace) {
+  /**
+   * The workspace this toolbox is on.
+   * @type {!Blockly.WorkspaceSvg}
+   * @protected
+   */
+  this.workspace_ = workspace;
+
+  /**
+   * The JSON describing the contents of this toolbox.
+   * @type {!Array<!Blockly.utils.toolbox.ToolboxItemDef>}
+   * @protected
+   */
+  this.toolboxDef_ = workspace.options.languageTree || [];
+
+  /**
+   * Whether the toolbox should be laid out horizontally.
+   * @type {boolean}
+   * @private
+   */
+  this.horizontalLayout_ = workspace.options.horizontalLayout;
+
+  /**
+   * The html container for the toolbox.
+   * @type {?Element}
+   */
+  this.HtmlDiv = null;
+
+  /**
+   * The html container for the contents of a toolbox.
+   * @type {?Element}
+   * @protected
+   */
+  this.contentsDiv_ = null;
+
+  /**
+   * The list of items in the toolbox.
+   * @type {!Array<!Blockly.ToolboxItem>}
+   * @protected
+   */
+  this.contents_ = [];
+
+  /**
+   * The width of the toolbox.
+   * @type {number}
+   * @protected
+   */
+  this.width_ = 0;
+
+  /**
+   * The height of the toolbox.
+   * @type {number}
+   * @protected
+   */
+  this.height_ = 0;
+
+  /**
+   * Is RTL vs LTR.
+   * @type {boolean}
+   */
+  this.RTL = workspace.options.RTL;
+
+  /**
+   * The flyout for the toolbox.
+   * @type {?Blockly.Flyout}
+   * @private
+   */
+  this.flyout_ = null;
+
+  /**
+   * The id of the toolbox item as the key and the toolbox item as the value.
+   * @type {!Object<string, Blockly.ToolboxItem>}
+   * @protected
+   */
+  this.contentIds_ = {};
+
+  /**
+   * Position of the toolbox and flyout relative to the workspace.
+   * @type {number}
+   */
+  this.toolboxPosition = workspace.options.toolboxPosition;
+
+  /**
+   * The currently selected item.
+   * @type {?Blockly.SelectableToolboxItem}
+   * @protected
+   */
+  this.selectedItem_ = null;
+
+  /**
+   * Array holding info needed to unbind events.
+   * Used for disposing.
+   * Ex: [[node, name, func], [node, name, func]].
+   * @type {!Array<!Array<?>>}
+   * @protected
+   */
+  this.boundEvents_ = [];
+};
+
+/**
+ * Initializes the toolbox
+ * @public
+ */
+Blockly.Toolbox.prototype.init = function() {
+  var workspace = this.workspace_;
+  var svg = workspace.getParentSvg();
+
+  this.flyout_ = this.createFlyout_();
+
+  this.HtmlDiv = this.createDom_(this.workspace_);
+  this.render(this.toolboxDef_);
+  var themeManager = workspace.getThemeManager();
+  themeManager.subscribe(this.HtmlDiv, 'toolboxBackgroundColour',
+      'background-color');
+  themeManager.subscribe(this.HtmlDiv, 'toolboxForegroundColour', 'color');
+
+  // Insert the flyout after the workspace.
+  Blockly.utils.dom.insertAfter(this.flyout_.createDom('svg'), svg);
```

Comment:
My gut feeling is that the flyout should be initialized--or at least its DOM should be created--before rendering the toolbox.  Thoughts?

Reply:
I don't have a lot of opinions on this. I'm not entirely sure what the benefit would be for creating the dom before the toolbox?

Reply:
I'm thinking of the scratch-style one, where the flyout is always open next to the toolbox. In that case I think when you render the toolbox it assumes the flyout dom exists.

Reply:
Ahh makes sense. Ok I moved up the createDom and init since I think you need both in order to call flyout.show_ in the render method.